File: examples/clojure/smoke.clj
;; Copyright (c) Andreas Flakstad and Vev contributors
;; SPDX-License-Identifier: EPL-2.0

(ns smoke
  (:require [vev.core :as vev]))

(defn delete-sqlite-files! [path]
  (doseq [suffix ["" "-wal" "-shm"]]
    (java.nio.file.Files/deleteIfExists
     (java.nio.file.Path/of (str path suffix) (make-array String 0)))))

(defn -main [& args]
  (when-not (= 1 (count args))
    (throw (ex-info "usage: smoke <path-to-native-library>" {})))
  (let [lib-path (first args)]
    (with-open [closed-conn (vev/create-conn lib-path)]
      (vev/transact closed-conn [{:db/id 10 :user/name "Closed"}])
      (let [closed-db (vev/db closed-conn)]
        (.close closed-conn)
        (when-not (= #{["Closed"]}
                     (vev/q closed-db
                            '[:find ?name :where [?e :user/name ?name]]))
          (throw (ex-info "DB value did not survive connection close" {})))))

    (with-open [conn (vev/create-conn lib-path)]
      (let [tx (vev/transact conn
                             [{:db/id 1 :user/name "Ada" :user/email "ada@example.com"}
                              {:db/id 2 :user/name "Grace" :user/email "grace@example.com"}])]
        (println "tx:" tx)
        (when-not (and (:ok tx) (integer? (:tx tx)) (map? (:tempids tx)))
          (throw (ex-info "unexpected transaction report" {:report tx}))))

      (let [seen (atom [])]
        (with-open [_listener (vev/listen conn :smoke-listener
                                          #(swap! seen conj (:tx-data %)))]
          (vev/transact conn [[:db/add 1 :user/listener "heard"]])
          (when-not (= 1 (count @seen))
            (throw (ex-info "listener did not observe successful transaction"
                            {:seen @seen}))))
        (vev/transact conn [[:db/add 1 :user/listener "after-unlisten"]])
        (when-not (= 1 (count @seen))
          (throw (ex-info "listener observed transaction after close"
                          {:seen @seen}))))

      (let [db (vev/db conn)
            names (vev/q
                   db
                   '[:find ?name
                     :in $ [?email ...]
                     :where [?e :user/email ?email]
                     [?e :user/name ?name]]
                   ["ada@example.com" "grace@example.com"])]
        (println "input-collection:" names)
        (when-not (= #{["Ada"] ["Grace"]} names)
          (throw (ex-info "unexpected collection query output" {:rows names}))))

      (let [db (vev/db conn)
            requested (vev/query
                       {:query '[:find ?name
                                 :in $ ?email
                                 :where [?e :user/email ?email]
                                 [?e :user/name ?name]]
                        :args [db "ada@example.com"]})]
        (println "query-map:" requested)
        (when-not (= #{["Ada"]} requested)
          (throw (ex-info "unexpected query map output" {:rows requested}))))

      (let [db (vev/db conn)
            keyed (vev/query
                   {:query '[:find ?name ?email
                             :keys name email
                             :in $ ?email
                             :where [?e :user/email ?email]
                             [?e :user/name ?name]]
                    :args [db "ada@example.com"]})]
        (println "query-map keys:" keyed)
        (when-not (= #{{:name "Ada" :email "ada@example.com"}} keyed)
          (throw (ex-info "unexpected query map keyed output" {:rows keyed}))))

      (vev/transact conn
                    [[:db/add 90 :db/ident :user/email]
                     [:db/add 90 :db/unique :db.unique/identity]])

      (vev/transact conn
                    [[:db/add 91 :db/ident :user/status]
                     [:db/add 91 :db/unique :db.unique/identity]
                     [:db/add 92 :db/ident :user/code]
                     [:db/add 92 :db/unique :db.unique/identity]
                     [:db/add 1 :user/status :active]
                     [:db/add 2 :user/status :inactive]
                     [:db/add 1 :user/code 1001]
                     [:db/add 2 :user/code 1002]])

      (vev/transact conn
                    [[:db/add 100 :db/ident :user/friend]
                     [:db/add 100 :db/valueType :db.type/ref]
                     [:db/add 1 :user/friend 2]])

      (let [db (vev/db conn)]
        (with-open [ada (vev/entity db 1)
                    friend (vev/entity-ref ada :user/friend)
                    lookup (vev/entity db [:user/email "ada@example.com"])]
          (when-not (and (vev/entity-found? ada)
                         (= 1 (vev/entity-id ada))
                         (= "Ada" (:user/name ada))
                         (= ["Ada"] (vev/entity-values ada :user/name))
                         (= 2 (vev/entity-id friend))
                         (= "Grace" (:user/name friend))
                         (= [2] (vev/entity-refs ada :user/friend))
                         (= 1 (vev/entity-id lookup))
                         (= "Ada" (:user/name (vev/touch ada))))
            (throw (ex-info "unexpected entity view output"
                            {:entity (vev/touch ada)})))))

      (with-open [fns (vev/tx-fns conn {:user/set-name
                                        (fn [db e name]
                                          [[:db/add e :user/nickname name]])})]
        (vev/transact conn [[:db/add 101 :db/ident :user/set-name]])
        (let [tx-fn-report (vev/transact conn [[:user/set-name 6 "Edsger"]] fns)]
          (when-not (:ok tx-fn-report)
            (throw (ex-info "transaction function failed" {:report tx-fn-report})))
          (when-not (= #{["Edsger"]}
                       (vev/q '[:find ?name :where [6 :user/nickname ?name]]
                              (vev/db conn)))
            (throw (ex-info "transaction function did not apply tx-data" {})))))

      (with-open [email-query (vev/prepare conn
                                           '[:find ?e ?email
                                             :in $ ?needle
                                             :where [?e :user/email ?email]
                                             [(= ?email ?needle)]])]
        (let [prepared-ast (vev/prepared-edn email-query)]
          (when-not (and (:ok prepared-ast)
                         (seq (:clauses prepared-ast))
                         (seq (:input-specs prepared-ast)))
            (throw (ex-info "prepared query AST did not expose parser keys"
                            {:prepared prepared-ast}))))
        (let [clause-ast (vev/parse-clause conn '[?e :user/email ?email])]
          (when-not (and (:ok clause-ast)
                         (seq (:clauses clause-ast)))
            (throw (ex-info "parse-clause AST did not expose parser keys"
                            {:clause clause-ast}))))

        (let [db (vev/db conn)
              rows (vev/q db email-query "grace@example.com")]
          (println "prepared rows:" rows)
          (when-not (= #{[2 "grace@example.com"]} rows)
            (throw (ex-info "unexpected prepared rows" {:rows rows}))))

        (with-open [rule-query (vev/prepare conn
                                            '[:find ?name
                                              :in $ % ?email
                                              :where (email-name ?email ?name)])]
          (let [db (vev/db conn)
                rules '[[(email-name ?email ?name)
                         [?e :user/email ?email]
                         [?e :user/name ?name]]]
                rows (vev/q rule-query db rules "ada@example.com")]
            (println "prepared rule rows:" rows)
            (when-not (= #{["Ada"]} rows)
              (throw (ex-info "unexpected prepared rule rows" {:rows rows})))))

        (let [db (vev/db conn)
              pulled (vev/pull db
                               [:user/name {:user/friend [:user/name]}]
                               1)]
          (println "pull:" pulled)
          (when-not (= {:user/name "Ada"
                        :user/friend {:user/name "Grace"}}
                       pulled)
            (throw (ex-info "unexpected pull result" {:pull pulled}))))

        (with-open [pull-column-query (vev/prepare conn
                                                   '[:find (pull ?e [:user/name {:user/friend [:user/name]}])
                                                     :where [?e :user/name "Ada"]])
                    db (vev/db conn)]
          (let [columns (vev/columns pull-column-query db)
                pulled (first (first (:columns columns)))]
            (println "pull column batch:" columns)
            (when-not (and (= [:value] (:kinds columns))
                           (= {:user/name "Ada"
                               :user/friend {:user/name "Grace"}}
                              pulled))
              (throw (ex-info "unexpected pull column batch"
                              {:columns columns})))))

        (let [db (vev/db conn)]
          (with-open [prepared-pattern (vev/prepare-pull-pattern db [:user/name])]
            (let [prepared-pattern-ast (vev/prepared-edn prepared-pattern)
                  lookup-pull (vev/pull db
                                        [:user/name]
                                        [:user/email "ada@example.com"])
                  prepared-lookup-pull (vev/pull db
                                                 prepared-pattern
                                                 [:user/email "ada@example.com"])
                  keyword-lookup-pull (vev/pull db
                                                [:user/name]
                                                [:user/status :active])
                  int-lookup-pull (vev/pull db
                                            [:user/name]
                                            [:user/code 1001])
                  many-pull (vev/pull-many db [:user/name] [1 2])]
              (println "lookup pull:" lookup-pull)
              (println "prepared lookup pull:" prepared-lookup-pull)
              (println "keyword lookup pull:" keyword-lookup-pull)
              (println "int lookup pull:" int-lookup-pull)
              (println "pull many:" many-pull)
              (when-not (and (:ok prepared-pattern-ast)
                             (seq (:pattern prepared-pattern-ast))
                             (= :user/name (:attr (first (:pattern prepared-pattern-ast)))))
                (throw (ex-info "prepared pull pattern AST did not expose parser keys"
                                {:prepared prepared-pattern-ast})))
              (when-not (= {:user/name "Ada"} lookup-pull)
                (throw (ex-info "unexpected lookup-ref pull" {:pull lookup-pull})))
              (when-not (= {:user/name "Ada"} prepared-lookup-pull)
                (throw (ex-info "unexpected prepared lookup-ref pull" {:pull prepared-lookup-pull})))
              (when-not (= {:user/name "Ada"} keyword-lookup-pull)
                (throw (ex-info "unexpected keyword lookup-ref pull" {:pull keyword-lookup-pull})))
              (when-not (= {:user/name "Ada"} int-lookup-pull)
                (throw (ex-info "unexpected int lookup-ref pull" {:pull int-lookup-pull})))
              (when-not (= #{"Ada" "Grace"} (set (map :user/name many-pull)))
                (throw (ex-info "unexpected pull-many" {:pull many-pull}))))))

        (with-open [all-names (vev/prepare conn
                                           '[:find ?name
                                             :where [?e :user/name ?name]])
                    db (vev/db conn)]
          (let [columns (vev/columns all-names db)]
            (println "column batch:" columns)
            (when-not (and (= [:string] (:kinds columns))
                           (= #{"Ada" "Grace"} (set (first (:columns columns)))))
              (throw (ex-info "unexpected column batch" {:columns columns})))))

        (let [db (vev/db conn)
              immutable-report (vev/with db [{:db/id 4 :user/name "Barbara"}])
              immutable-report-with-dbs (vev/with-report db [{:db/id 4 :user/name "Barbara"}])
              immutable-next (vev/db-with db [{:db/id 4 :user/name "Barbara"}])]
          (println "with tx:" immutable-report)
          (when-not (and (:ok immutable-report)
                         (= #{} (vev/q db
                                       '[:find ?e
                                         :where [?e :user/name "Barbara"]])))
            (throw (ex-info "immutable with mutated source DB" {:report immutable-report})))
          (try
            (when-not (and (= #{} (vev/q (:db-before immutable-report-with-dbs)
                                         '[:find ?e
                                           :where [?e :user/name "Barbara"]]))
                           (= #{[4]} (vev/q (:db-after immutable-report-with-dbs)
                                            '[:find ?e
                                              :where [?e :user/name "Barbara"]])))
              (throw (ex-info "with-report db-before/db-after mismatch"
                              {:report immutable-report-with-dbs})))
            (finally
              (.close (:db-before immutable-report-with-dbs))
              (.close (:db-after immutable-report-with-dbs))))
          (when-not (= #{[4]}
                       (vev/q immutable-next
                              '[:find ?e
                                :where [?e :user/name "Barbara"]]))
            (throw (ex-info "db-with did not produce expected DB" {})))
          (with-open [derived-conn (vev/conn-from-db immutable-next)]
            (vev/transact derived-conn [{:db/id 5 :user/name "Dorothy"}])
            (when-not (and (= #{[4]}
                              (vev/q (vev/db derived-conn)
                                     '[:find ?e
                                       :where [?e :user/name "Barbara"]]))
                           (= #{[5]}
                              (vev/q (vev/db derived-conn)
                                     '[:find ?e
                                       :where [?e :user/name "Dorothy"]])))
              (throw (ex-info "conn-from-db did not initialize from DB value" {})))))

        (with-open [all-emails (vev/prepare conn
                                            '[:find ?e ?email
                                              :where [?e :user/email ?email]])
                    snapshot (vev/db conn)]
          (vev/transact conn
                        [{:db/id 3 :user/name "Alan" :user/email "alan@example.com"}])
          (let [current (vev/db conn)
                current-rows (count (vev/q current all-emails))
                snapshot-rows (count (vev/q snapshot all-emails))]
            (println "current-db rows:" current-rows)
            (println "snapshot-db rows:" snapshot-rows)
            (when-not (and (= 3 current-rows) (= 2 snapshot-rows))
              (throw (ex-info "unexpected snapshot row counts"
                              {:current current-rows :snapshot snapshot-rows})))))))

    (let [sqlite-path "tmp.vev.clojure.sqlite"]
      (delete-sqlite-files! sqlite-path)
      (try
        (with-open [durable (vev/connect lib-path sqlite-path)]
          (when (not= {:backend :sqlite :path sqlite-path :basis-t 0 :tx-count 0 :tx-ids []} (vev/connection-info durable))
            (throw (ex-info "unexpected durable connection metadata" {})))
          (let [tx (vev/transact durable
                                 [{:db/id 1
                                   :user/name "Durable Ada"
                                   :user/email "durable-ada@example.com"}])]
            (when-not (:ok tx)
              (throw (ex-info "unexpected SQLite transaction report" {:report tx}))))
          (when (not= 1 (:basis-t (vev/connection-info durable)))
            (throw (ex-info "unexpected durable basis after first tx" {})))
          (when (not= 1 (:tx-count (vev/connection-info durable)))
            (throw (ex-info "unexpected durable tx count after first tx" {})))
          (when (not= [1] (:tx-ids (vev/connection-info durable)))
            (throw (ex-info "unexpected durable tx ids after first tx" {})))
          (with-open [bulk-a (vev/tx-builder durable 1)
                      bulk-b (vev/tx-builder durable 1)]
            (vev/tx-add! bulk-a 2 :user/name "Durable Grace")
            (vev/tx-add! bulk-b 3 :user/name "Durable Hedy")
            (let [tx (vev/transact-bulk durable [bulk-a bulk-b])]
              (when-not (:ok tx)
                (throw (ex-info "unexpected durable bulk builder report"
                                {:report tx})))))
          (when (not= 2 (:basis-t (vev/connection-info durable)))
            (throw (ex-info "unexpected durable basis after bulk builder tx" {})))
          (when (not= 2 (:tx-count (vev/connection-info durable)))
            (throw (ex-info "unexpected durable tx count after bulk builder tx" {})))
          (when (not= [1 2] (:tx-ids (vev/connection-info durable)))
            (throw (ex-info "unexpected durable tx ids after bulk builder tx" {})))
          (with-open [logical-a (vev/tx-builder durable 1)
                      logical-b (vev/tx-builder durable 1)]
            (vev/tx-add! logical-a 4 :user/name "Durable Ada")
            (vev/tx-add! logical-b 5 :user/name "Durable Dorothy")
            (let [reports (vev/transact-logical-bulk durable [logical-a logical-b])]
              (when-not (and (= 2 (count reports))
                             (every? :ok reports))
                (throw (ex-info "unexpected durable logical group reports"
                                {:reports reports})))))
          (let [reports (vev/transact-logical-bulk durable [])]
            (when-not (empty? reports)
              (throw (ex-info "unexpected empty logical group reports"
                              {:reports reports}))))
          (let [reports (vev/transact-logical durable
                                              [[{:db/id 6
                                                 :user/name "Durable Katherine"}]
                                               "[{:db/id 7 :user/name \"Durable Mary\"}]"])]
            (when-not (and (= 2 (count reports))
                           (every? :ok reports))
              (throw (ex-info "unexpected durable logical EDN group reports"
                              {:reports reports}))))
          (when (not= 6 (:basis-t (vev/connection-info durable)))
            (throw (ex-info "unexpected durable basis after logical group tx" {})))
          (when (not= 6 (:tx-count (vev/connection-info durable)))
            (throw (ex-info "unexpected durable tx count after logical group tx" {})))
          (when (not= [1 2 3 4 5 6] (:tx-ids (vev/connection-info durable)))
            (throw (ex-info "unexpected durable tx ids after logical group tx" {})))
          (when-not (vev/compact-indexes! durable)
            (throw (ex-info "durable index compaction failed" {})))
          (with-open [db (vev/db durable)
                      all-emails (vev/prepare durable
                                              '[:find ?e ?email
                                                :where [?e :user/email ?email]])]
            (let [rows (vev/q db all-emails)]
              (println "sqlite-live rows:" rows)
              (when-not (= 1 (count rows))
                (throw (ex-info "unexpected SQLite live row count" {:rows rows}))))))
        (with-open [durable (vev/connect lib-path sqlite-path)
                    db (vev/db durable)
                    all-emails (vev/prepare durable
                                            '[:find ?e ?email
                                              :where [?e :user/email ?email]])]
          (when (not= 6 (:basis-t (vev/connection-info durable)))
            (throw (ex-info "unexpected reopened durable basis" {})))
          (when (not= 6 (:tx-count (vev/connection-info durable)))
            (throw (ex-info "unexpected reopened durable tx count" {})))
          (when (not= [1 2 3 4 5 6] (:tx-ids (vev/connection-info durable)))
            (throw (ex-info "unexpected reopened durable tx ids" {})))
          (let [rows (vev/q db all-emails)]
            (println "sqlite-reopened rows:" rows)
            (when-not (= 1 (count rows))
              (throw (ex-info "unexpected SQLite reopened row count" {:rows rows})))))
        (finally
          (delete-sqlite-files! sqlite-path))))))

(apply -main *command-line-args*)

File: clients/node/smoke.js
// Copyright (c) Andreas Flakstad and Vev contributors
// SPDX-License-Identifier: EPL-2.0

"use strict";

const vev = require("./vev");
const fs = require("fs");

function mustContain(label, text, ...needles) {
  for (const needle of needles) {
    if (!text.includes(needle)) {
      throw new Error(`${label} missing ${needle} in ${text}`);
    }
  }
}

function removeSqliteFiles(path) {
  for (const suffix of ["", "-wal", "-shm"]) {
    try {
      fs.unlinkSync(`${path}${suffix}`);
    } catch (error) {
      if (error.code !== "ENOENT") {
        throw error;
      }
    }
  }
}

const conn = vev.createConn();

const tx = conn.transact(`
  [{:db/id 1 :user/name "Ada" :user/email "ada@example.com"}
   {:db/id 2 :user/name "Grace" :user/email "grace@example.com"}]
`);
console.log("tx:", tx);
mustContain("tx", tx, ":ok true");

const collection = conn.queryText(`
  [:find ?name
   :in $ [?email ...]
   :where [?e :user/email ?email]
          [?e :user/name ?name]]
`, `[["ada@example.com" "grace@example.com"]]`);
console.log("input-collection:", collection);
mustContain("collection query", collection, `"Ada"`, `"Grace"`);

const oneShotRows = conn.q('[:find ?name :where [?e :user/name ?name]]');
if (!(oneShotRows instanceof Set)) {
  throw new Error("one-shot relation query did not return a Set");
}
const oneShotNames = Array.from(oneShotRows, (row) => row[0]).sort();
if (JSON.stringify(oneShotNames) !== JSON.stringify(["Ada", "Grace"])) {
  throw new Error(`unexpected one-shot query rows: ${JSON.stringify(oneShotRows)}`);
}
const scalarName = conn.q('[:find ?name . :where [1 :user/name ?name]]');
if (scalarName !== "Ada") {
  throw new Error(`unexpected scalar query result: ${scalarName}`);
}

const query = conn.prepare(`
  [:find ?e ?email
   :in $ ?needle
   :where [?e :user/email ?email]
          [(= ?email ?needle)]]
`);

mustContain("prepared AST", query.edn(), ":clauses", ":input-specs");

const prepared = query.query(conn, `["grace@example.com"]`);
console.log("prepared:", prepared);
mustContain("prepared query", prepared, "2", `"grace@example.com"`);

const rows = query.rows(conn, `["grace@example.com"]`);
console.log("typed rows:", rows);
if (rows.length !== 1 || rows[0][0].id !== 2 || rows[0][1] !== "grace@example.com") {
  throw new Error(`unexpected typed rows: ${JSON.stringify(rows)}`);
}

conn.transact(`
  [[:db/add 90 :db/ident :user/email]
   [:db/add 90 :db/unique :db.unique/identity]]
`);

conn.transact(`
  [[:db/add 100 :db/ident :user/friend]
   [:db/add 100 :db/valueType :db.type/ref]
   [:db/add 1 :user/friend 2]]
`);

const pullQuery = conn.prepare(`
  [:find (pull ?e [:user/name {:user/friend [:user/name]}])
   :where [?e :user/name "Ada"]]
`);
const pulled = pullQuery.rows(conn, "[]")[0][0];
console.log("pull:", pulled);
if (pulled[":user/name"] !== "Ada" || pulled[":user/friend"][":user/name"] !== "Grace") {
  throw new Error(`unexpected pull: ${JSON.stringify(pulled)}`);
}
pullQuery.close();

const snapshot = conn.db();
const directPull = snapshot.pull("[:user/name {:user/friend [:user/name]}]", 1);
console.log("direct pull:", directPull);
if (directPull[":user/name"] !== "Ada" || directPull[":user/friend"][":user/name"] !== "Grace") {
  throw new Error(`unexpected direct pull: ${JSON.stringify(directPull)}`);
}
const lookupPull = snapshot.pullLookupRefString("[:user/name]", ":user/email", "ada@example.com");
if (lookupPull[":user/name"] !== "Ada") {
  throw new Error(`unexpected lookup pull: ${JSON.stringify(lookupPull)}`);
}
const manyPull = snapshot.pullMany("[:user/name]", [1, 2]);
const manyNames = manyPull.map((item) => item[":user/name"]).sort();
if (JSON.stringify(manyNames) !== JSON.stringify(["Ada", "Grace"])) {
  throw new Error(`unexpected pull-many: ${JSON.stringify(manyPull)}`);
}

conn.transact(`[{:db/id 3 :user/name "Alan" :user/email "alan@example.com"}]`);

const current = query.query(conn, `["alan@example.com"]`);
const old = snapshot.query(query, `["alan@example.com"]`);
console.log("current-db:", current);
console.log("snapshot-db:", old);
mustContain("current DB query", current, "3", `"alan@example.com"`);
if (old.includes("alan@example.com")) {
  throw new Error("snapshot unexpectedly observed later transaction");
}

const currentRows = query.rows(conn, `["alan@example.com"]`);
const oldRows = snapshot.rows(query, `["alan@example.com"]`);
if (currentRows.length !== 1 || oldRows.length !== 0) {
  throw new Error("unexpected typed snapshot row counts");
}

const withReport = snapshot.withReport(
  `[{:db/id 4 :user/name "Barbara" :user/email "barbara@example.com"}]`,
);
if (!withReport.edn.includes(":ok true")) {
  throw new Error(`unexpected with report: ${withReport.edn}`);
}
const reportBeforeRows = withReport.dbBefore.rows(query, `["barbara@example.com"]`);
const reportAfterRows = withReport.dbAfter.rows(query, `["barbara@example.com"]`);
if (reportBeforeRows.length !== 0 || reportAfterRows.length !== 1) {
  throw new Error("unexpected with report DB rows");
}
withReport.dbBefore.close();
withReport.dbAfter.close();
snapshot.close();

const sqlitePath = "tmp.vev.node.sqlite";
removeSqliteFiles(sqlitePath);
try {
  const durable = vev.connect(sqlitePath);
  if (
    durable.backend() !== "sqlite" ||
    durable.path() !== sqlitePath ||
    durable.basisT() !== 0 ||
    durable.txCount() !== 0
  ) {
    throw new Error("unexpected durable metadata");
  }
  durable.transact(`[{:db/id 1 :user/name "Durable Ada" :user/email "durable-ada@example.com"}]`);
  const firstBasis = durable.basisT();
  if (firstBasis === 0 || durable.txCount() !== 1) {
    throw new Error("unexpected durable metadata after transact");
  }
  const durableQuery = conn.prepare(`[:find ?e ?email :in $ ?needle :where [?e :user/email ?email] [(= ?email ?needle)]]`);
  const durableDb = durable.db();
  const durableRows = durableDb.rows(durableQuery, `["durable-ada@example.com"]`);
  if (durableRows.length !== 1 || durableRows[0][0].id !== 1) {
    throw new Error(`unexpected durable rows: ${JSON.stringify(durableRows)}`);
  }
  durableDb.close();
  durableQuery.close();
  durable.close();
} finally {
  removeSqliteFiles(sqlitePath);
}

query.close();
conn.close();
